### ISSUE TITLE
write() inside an @onchange does not work as expected

### ISSUE THREAD

**Issue by Yajo** [odoo/odoo#7417]
When inside an onchange method you call `write()`, it does not update the UI. However, assigning the attributes one by one works.

Suppose this code. It's buggy but it's only for demonstration purposes:

```
class Partner(models.Model):
  @api.one
  @api.onchange("name")
  def _onchange_name(self):
    parts = self.name.split(None, 3)
    while len(parts) < 3:
      parts.append(False)
    self.firstname, self.lastname, self.lastname2 = parts[2], parts[0], parts[1]

  @api.one
  @api.constrains("firstname", "lastname", "lastname2")
  def _check_names(self):
    if not (self.firstname or self.lastname or self.lastname2):
      raise orm_exception()
```

This code will trigger 3 times `_check_names`, which is not only inefficient but also buggy: If somebody writes a name like "One Two Three", it will work, but if somebody writes "One" it will except because the first time that `_check_names` is executed, all names are `False`.

The logical step to avoid that is to replace the last line of `_onchange_name` with:

    self.write({"firstname": parts[2], "lastname": parts[0], "lastname2": parts[1]})

... which should trigger only once `_check_names` with all fields modified. But that will never trigger `_check_names` because it's being called from an onchange method. **That's the bug**.

The only current workaround possible is replace the last line of `_onchange_name` with:

    fields = ("lastname", "lastname2", "firstname")
    # First, write the fields with data
    for field in range(3):
        if parts[field]:
            setattr(self, fields[field], parts[field])
    # Last, write the empty fields
    for field in range(3):
        if not parts[field]:
            setattr(self, fields[field], parts[field])

... which is *very ugly*.

**Comment by andreparames** [comment]
write() is only for actually putting the data on the database. For this, you should use update():

  self.update({"firstname": parts[2], "lastname": parts[0], "lastname2": parts[1]})

**Comment by Yajo** [comment]
Nice! Then the bug is that [it's not documented](https://www.odoo.com/documentation/8.0/reference/orm.html#common-orm-methods).

**Comment by Yenthe666** [comment]
@mart-e perhaps this is something for you? :smile:

**Comment by mart-e** [comment]
No look, it's [documented here](https://docs.python.org/2.7/library/stdtypes.html#dict.update) :wink: 
The `update` method on a recordset is what you would expect on a dictionary.

**Comment by Yajo** [comment]
Yes, but it is not a dict. You are not inheriting dict to declare the recordset object afaik... Or does it have items, iteritems, keys, values, iterkeys, itervalues, get, pop, etc.?

The problem is not that it does what you expect, the problem is that you don't know what to expect because you don't even know if it exists because it is not documented.

**Comment by dreispt** [comment]
@mart-e Is using the `update` equivalent to assigning each field value individually?
If so, it should share it's limitations: it only works on singletons and performs a database write for each written field.

**Comment by Yajo** [comment]
Hopefully that will not happen from v10 since they merged https://github.com/odoo/odoo/pull/9210.

The main difference is that update works fine inside onchanges (AKA, does not writes).

**Comment by dreispt** [comment]
@Yajo Said that way, sounds like a problem to be fixed with in ` write` method in the first place...

**Comment by Yajo** [comment]
Unless something changes, Odoo has right now different methods for working with cache records and database records: new=create, update=write.

Cached versions should not write to database until cache is consolidated.

Why do these differences exist when simply create and write (and unlink, why not?) could detect they are under a `with self.env.do_in_onchange():` block and avoid db writes if so? That, dear friend, is a very good question. However I feel it's away from this bug's title.

**Comment by dreispt** [comment]
It would be great if @rco-odoo could comment on that.

**Comment by dreispt** [comment]
(or @odony )

**Comment by Yenthe666** [comment]
@mart-e can you have a final look, give feedback again & close this topic off? 😄

**Comment by arveitch** [comment]
I found this searching for documentation on update(). Unfortunately it is becoming documented on the web with wrong information, which cost me quite a lot of time.

Adding a sentence to the ORM documentation explaining update() would be very helpful.

**Comment by pedrobaeza** [comment]
@rco-odoo maybe you can consider to add something in the documentation for explaining the correct use of update?

**Comment by jjscarafia** [comment]
:+1:  for clarifying recommendations and use case for .new vs .create and .write vs .update
BR!

**Comment by dreispt** [comment]
I believe that @odony mentioned somwhere that this is an internal implementation detail, not part of the official API, and so should be be used/relied on.

**Comment by igormandic-sc** [comment]
@rco-odoo Apologies for bumping this, but can you provide some additional definitive information regarding this issue, and details on the functionality of the method in question?

@dreispt notes @odony mentioned that `update()` shouldn't be used since it's not part of the official API, is this true?

**Comment by C3POdoo** [comment]
Dear @yajo,

Thank you for your report but we are closing it due to inactivity.
We apology if we could not look at your request in time.
If your report still makes sense, don't hesitate to reopen a new one. We will try to check it as soon as possible.

_This is an automated message._
